### PR TITLE
Introduce evaluation on concept captions

### PR DESCRIPTION
**Pull request by gabrieletijunaityte**
## What does this PR do?

Implements top-k accuracy for concept captions, based on how many retrieved top-k entries are also top-k entries based on auxilary values.

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **test your PR locally** with `pytest` command?

### CODE REVIEW

File: data/s2bms/concept_captions/v1.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+[
+ {
+    "concept_caption": "Forested area",
+    "is_max": true,
+    "theta_k": 0.5,
+    "col": "aux_corine_frac_311"
```
Comment (vdplasthijs) [inline data/s2bms/concept_captions/v1.json]: Old column name?

Comment (vdplasthijs) [inline data/s2bms/concept_captions/v1.json]: This could be `aux_corine_frac_3`

File: data/s2bms/concept_captions/v1.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  },
+  {
+    "concept_caption": "Sparsely populated area",
+    "is_max": false,
+    "theta_k": 0.2,
+    "col": "aux_corine_frac_111"
```
Comment (vdplasthijs) [inline data/s2bms/concept_captions/v1.json]: this could be `aux_corine_frac_11`

File: src/models/text_alignment_model.py [hunk 3]
```diff
@@ -81,6 +87,21 @@ def __init__(
         # Freezing requested parts
         self.freezer()
 
+    def setup(self, stage: str) -> None:
+        self.concept_configs = self.trainer.datamodule.concept_configs
+        self.concepts = [c["concept_caption"] for c in self.concept_configs]
```
Comment (vdplasthijs) [inline src/models/text_alignment_model.py:94]: This assumes that the "concept_caption" entries in the json are all unique. That's fine I think, but might be worth having a check to assert this in the caption builder. (in case someone wants to test "forested area" against different subclasses of woodlands for example).

File: src/models/components/metrics/contrastive_validation.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+        aux_vals = aux_values.T
+
+        avr_scores = {k: [] for k in self.ks}
+        concept_scores = []
+        for i, configs in enumerate(self.concept_configs):
+            idx = configs["id"]
+            is_max = configs["is_max"]
+            theta_k = configs.get("theta_k")
+            if theta_k and aux_vals is not None:
```
Comment (vdplasthijs) [inline src/models/components/metrics/contrastive_validation.py]: I think an else statement is missing here. (because right after the if statment `aux_val` is used.

File: src/models/components/metrics/contrastive_validation.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                aux_val = aux_vals[idx]
+                theta_k = (
```
Comment (vdplasthijs) [inline src/models/components/metrics/contrastive_validation.py]: It would be better if this was called eg `k_dynamic` or `k_threshold` rather than reusing `theta_k`

File: src/models/components/metrics/contrastive_validation.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                avr_scores[k].append(v)
+
+        return_scores = {}
+        for k, v in avr_scores.items():
+            return_scores[f"avr_top-{k}"] = sum(v) / len(v)
+
+        return return_scores, concept_scores
+
+    @staticmethod
+    def topk_rank_agreement(gt_vals, pred_vals, ks, is_max=True, theta_k=None):
```
Comment (vdplasthijs) [inline src/models/components/metrics/contrastive_validation.py]: And then also change the name of  `theta_k` in this function.

File: src/models/components/metrics/contrastive_validation.py (hunk 5 above)
Comment (vdplasthijs) [reply src/models/components/metrics/contrastive_validation.py]: (maybe more precies would be `k_theta` that is defined by `theta`, but then we'd need to change both variable names.)

File: src/models/components/metrics/contrastive_validation.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        :param concept_configs: concept configurations containing details about min/max mode, which
+            aux_col to use as GT.
+        """
+        super().__init__()
+
+        self.concept_configs = concept_configs
+
+        self.ks = ks
+        if any("theta_k" in c for c in self.concept_configs):
+            self.ks.append("theta_k")
```
Comment (vdplasthijs) [inline src/models/components/metrics/contrastive_validation.py]: Change this name too

File: src/models/text_alignment_model.py [hunk 8]
```diff
@@ -122,18 +143,77 @@ def _step(self, batch: Dict[str, torch.Tensor], mode: str = "train") -> torch.Te
...
+        aux_vals = torch.cat([x["aux_vals"] for x in self.outputs_epoch_memory], dim=0)
+
+        # Rank on similarity
+        similarity = self.concept_similarities(eo_feats)
+
+        avr_scores, concept_scores = self.contrastive_val(similarity, aux_values=aux_vals)
+        # TODO pearson
+
+        self.log_dict(avr_scores)
+        for i, result in enumerate(concept_scores):
```
Comment (vdplasthijs) [inline src/models/text_alignment_model.py]: I think this fine, but perhaps clearer would be if `concept_scores` (returned by `ContrastiveValidation` would already be a dictionary of `{concepts[i]: score[i]}` (in pseudo-notation), if that makes sense? what do you think?

**Review by vdplasthijs — COMMENTED** [review]
Ok great work. Good thinking of where what component should go, I think that all makes sense and gives a clean implementation! I just have some minor suggestions, have a look? Thanks!!

File: src/models/text_alignment_model.py (hunk 3 above)
Comment (gabrieletijunaityte) [reply src/models/text_alignment_model.py:94]: But it's a list, so it allows duplicates. The only downside is that we would embed the same concept twice, but I don't expect it to happen often, so maybe it's not a big issue?

Comment (vdplasthijs) [reply src/models/text_alignment_model.py:94]: Sorry yes you're right. Ok yes that's fine, false alarm!

File: src/models/text_alignment_model.py (hunk 8 above)
Comment (gabrieletijunaityte) [reply src/models/text_alignment_model.py]: Do you mean concept id? Or the concept in textual form? And how would I return averages? Or no averages as they can be easily calculated?

Comment (vdplasthijs) [reply src/models/text_alignment_model.py]: Yes `id` is better if we allow for duplicates (which sounds good!). 

Averages can be calculated/returned in `ContrastiveValidation.forward()` as before right, they are calculated directly from `score`, independent of `concept_scores`?

**Review by vdplasthijs — APPROVED** [review]
Great thanks!!
